### ISSUE TITLE
Multiple Shared Memory Configs

### ISSUE THREAD

**Issue by blaazetech** [aerospike/aerospike-client-nodejs#191]
We have two clusters of AS with 4 nodes each and we are trying to use shared memory configuration, does the config supports to create only one hosts array and use same shared memory key or do we need to use separate keys and separate client objects

```
nodes in clusters

cluster 1
==========
192.168.0.1
192.168.0.2
192.168.0.3
192.168.0.4

cluster 2
==========
192.168.1.1
192.168.1.2
192.168.1.3
192.168.1.4
```

we are currently doing something like this
```
var aerospikeclient1 = Aerospike.client({sharedMemory: {key: 0xa5000000, takeoverThresholdSeconds: 2},hosts: [{ addr: "192.168.0.1", port: 3000 }]});
var aerospikeclient2 = Aerospike.client({sharedMemory: {key: 0xa6000000, takeoverThresholdSeconds: 2},hosts: [{ addr: "192.168.1.1", port: 3000 }]});
```

but what we are looking for is something like this
```
var aerospikeclientuniversal = Aerospike.client({sharedMemory: {key: 0xa5000000, takeoverThresholdSeconds: 2},hosts: [{ addr: "192.168.0.1", port: 3000 },{ addr: "192.168.1.1", port: 3000 }]});
```

please note we have different namespaces and different data on both clusters, we are trying to connect to both clusters using same client config object and same shm_key, is it possible to achieve or do we need to deploy separate configs for each cluster?

Also is it safe to use 0xa5000000 and 0xa6000000, does it have safe cushion memory area to store data,as we rely on millions of queries.

**Comment by blaazetech** [comment]
Post using separate connection with separate shm_key as above, we got issue of memory leakage or the memory cannot be allocated, and the whole process went down and restarted and eventually the whole application server crashes.
```
/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284
    throw new AerospikeError('Not connected.')
    ^

/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284
    throw new AerospikeError('Not connected.')
    ^

/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284
    throw new AerospikeError('Not connected.')
    ^
FATAL ERROR: Committing semi space failed. Allocation failed - process out of memory

<--- Last few GCs --->

 4000012 ms: Scavenge 26.8 (68.0) -> 26.1 (68.0) MB, 0.8 / 0 ms (+ 1.9 ms in 13 steps since last GC) [allocation failure].
 4102874 ms: Scavenge 27.0 (68.0) -> 26.2 (69.0) MB, 0.9 / 0 ms (+ 1.4 ms in 13 steps since last GC) [allocation failure].
 4208352 ms: Scavenge 27.1 (69.0) -> 26.4 (69.0) MB, 0.7 / 0 ms (+ 0.5 ms in 13 steps since last GC) [allocation failure].
 4315063 ms: Scavenge 27.3 (69.0) -> 26.5 (69.0) MB, 0.9 / 0 ms (+ 1.7 ms in 13 steps since last GC) [allocation failure].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x3bad6db4629 <JS Object>
    1: onconnection [native v8natives.js:1199] [pc=0x1c4ef6a6f890] (this=0xdf235d04df1 <a TCP with map 0x369e7b9d7469>)
    2: arguments adaptor frame: 2->0

==== Details ================================================

[1]: onconnection [native v8natives.js:1199] [pc=0x1c4ef6a6f890] (this=0xdf235d04df1 <a TCP with map 0x369e7b9d7469>) {
  // stack-allocated locals
  var arguments = 0x1450a0064...

error: Forever detected script was killed by signal: SIGABRT
error: Script restart attempt #1
AerospikeError: Aerospike Error
    at Client.query (/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284:11)
    at async.parallel.interquery (/home/ec2-user/prodcodes/index.js:3353:38)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:3671:13
    at replenish (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:884:21)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:888:13
    at eachOfLimit (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:915:26)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:920:20
    at eachOf (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:1052:9)
    at _parallel (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3670:9)
    at Object.parallelLimit [as parallel] (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3750:7)
AerospikeError: Aerospike Error
    at Client.query (/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284:11)
    at async.parallel.interquery (/home/ec2-user/prodcodes/index.js:3353:38)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:3671:13
    at replenish (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:884:21)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:888:13
    at eachOfLimit (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:915:26)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:920:20
    at eachOf (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:1052:9)
    at _parallel (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3670:9)
    at Object.parallelLimit [as parallel] (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3750:7)
AerospikeError: Aerospike Error
    at Client.query (/home/ec2-user/prodcodes/node_modules/aerospike/lib/client.js:1284:11)
    at async.parallel.interquery (/home/ec2-user/prodcodes/index.js:3353:38)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:3671:13
    at replenish (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:884:21)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:888:13
    at eachOfLimit (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:915:26)
    at /home/ec2-user/prodcodes/node_modules/async/dist/async.js:920:20
    at eachOf (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:1052:9)
    at _parallel (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3670:9)
    at Object.parallelLimit [as parallel] (/home/ec2-user/prodcodes/node_modules/async/dist/async.js:3750:7)
Connection to Aerospike cluster1 succeeded!
Connection to Aerospike cluster2 succeeded!
Listening on 80
```

**Comment by blaazetech** [comment]
Node version - v4.6.0
Aerospike Node.js Client - v2.5.2
Aerospike - Aerospike Community Edition build 3.7.5.1

**Comment by jhecking** [comment]
Each client instance connects to a single cluster. So you do need multiple client instances to connect to two separate clusters.

The shared memory setup is useful if you have multiple clients on the same machine but in different processes that all connect to the same cluster. E.g. if you are using the `cluster` module. The size of the shared memory segment is fairly small (typically less than 1 MB) and static, i.e. it is initialized once and does not grow over time. The size only depends on the number of cluster nodes and number of namespaces you use. It does not depend on the total amount of data you process or store in the cluster. More info is available in the [C client docs on shared memory usage](http://www.aerospike.com/docs/client/c/usage/shm.html). (But I assume you have read this already.)

We recently discovered a memory lead in Batch Reads with the Aerospike C client. The fix for that was included in v2.5 of the Node.js client (via upgrade of the included C client library).

**Comment by blaazetech** [comment]
Thanks for the information about the shared memory, but we are still facing issues with memory, as you said we are already on following version
```
Node version - v4.6.0
Aerospike Node.js Client - v2.5.2
Aerospike - Aerospike Community Edition build 3.7.5.1
```

**Comment by blaazetech** [comment]
upgrading to 3.12.1 and latest version of client had same issue, but got it solved i guess with 3.12.1.1 as we didn't see the same error again.
